synapse/handlers/room_member.py

```diff
@@ -178,6 +179,59 @@ async def ratelimit_invite(
 
         await self._invites_per_user_limiter.ratelimit(requester, invitee_user_id)
 
+    async def _can_join_restricted_room(
+        self, state_ids: StateMap[str], room_version: RoomVersion, user_id: str
+    ) -> bool:
+        """
+        Check whether a user can join a restricted room.
+
+        Args:
+            state_ids: The state of the room as it currently is.
+            room_version: The room version of the room being joined.
+            user_id: The user joining the room.
+
+        Returns:
+            True if the user can join the room, false otherwise.
+        """
+        # This only applies to room versions which support the new join rule.
+        if not room_version.msc3083_join_rules:
+            return True
+
+        # If there's no join rule, then it defaults to public (so this doesn't apply).
+        join_rules_event_id = state_ids.get((EventTypes.JoinRules, ""), None)
+        if not join_rules_event_id:
+            return True
+
+        # If the join rule is not restricted, this doesn't apply.
+        join_rules_event = await self.store.get_event(join_rules_event_id)
+        if join_rules_event.content.get("join_rule") != JoinRules.MSC3083_RESTRICTED:
+            return True
+
+        # If allowed is of the wrong form, then ignore it (and treat the room as public).
```

Comment:
> (and treat the room as public).

this feels like a questionable decision - we should fail towards more security.

Reply:
I seem to have swapped this from the original MSC2962 phrasing of:

> If the `allow` key is an empty list (or not a list at all), no users are allowed to join without an invite.

To the following in MSC3083:

> If the `allow` key is an empty list (or not a list at all), then the room reverts to standard `public` join rules.

I don't really recall *why* I did that, but I'm guessing it was due to the phrasing of how `restricted` differs from other join rules:

> the same as `public` from the perspective of the auth rules, but with the additional caveat that servers are expected to check the `allow` rules before generating a `join` event (whether for a local or a remote user).

I agree though that this seems quite questionable. I'm going to change this back to the original MSC2962 phrasing.

Reply:
I made this change (and updated the MSC).